storage/remote/dial_context_test.go

```diff
@@ -0,0 +1,151 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"errors"
+	"net"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	testNetwork         = "tcp"
+	testAddrWithoutPort = "this-is-my-addr.without-port"
+	testAddrWithPort    = "this-is-my-addr.without-port:123"
+	testPort            = "123"
+	ip1                 = "1.2.3.4"
+	ip2                 = "5.6.7.8"
+	ip3                 = "9.0.1.2"
+)
+
+var (
+	errMockLookupHost        = errors.New("this is a mocked error")
+	testLookupResult         = []string{ip1, ip2, ip3}
+	testLookupResultWithPort = []string{net.JoinHostPort(ip1, testPort), net.JoinHostPort(ip2, testPort), net.JoinHostPort(ip3, testPort)}
+)
+
+type mockDialContext struct {
+	mock.Mock
+
+	addrFrequencyMu sync.Mutex
+	addrFrequency   map[string]int
+}
+
+func newMockDialContext(acceptableAddresses []string) *mockDialContext {
+	m := &mockDialContext{
+		addrFrequencyMu: sync.Mutex{},
+		addrFrequency:   make(map[string]int),
+	}
+	for _, acceptableAddr := range acceptableAddresses {
+		m.On("dialContext", mock.Anything, mock.Anything, acceptableAddr).Return(nil, nil)
+	}
+	return m
+}
+
+func (dc *mockDialContext) dialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	dc.addrFrequencyMu.Lock()
+	defer dc.addrFrequencyMu.Unlock()
+	args := dc.MethodCalled("dialContext", ctx, network, addr)
+	dc.addrFrequency[addr]++
+	return nil, args.Error(1)
+}
+
+func (dc *mockDialContext) getCount(addr string) int {
+	dc.addrFrequencyMu.Lock()
+	defer dc.addrFrequencyMu.Unlock()
+	return dc.addrFrequency[addr]
+}
+
+type mockedLookupHost struct {
+	withErr bool
+	result  []string
+}
+
+func (lh *mockedLookupHost) LookupHost(context.Context, string) ([]string, error) {
+	if lh.withErr {
+		return nil, errMockLookupHost
+	}
+	return lh.result, nil
+}
+
+func TestDialContextWithRandomConnections(t *testing.T) {
+	numberOfRuns := 2 * len(testLookupResult)
+	var mdc *mockDialContext
+	testCases := map[string]struct {
+		addr                              string
+		setup                             func() dialContextWithRoundRobinDNS
+		check                             func()
+		withLookupHostErr                 bool
+		expectedDefaultContextArguments   []string
+		unexpectedDefaultContextArguments []string
+	}{
+		"if address contains no port call default DealContext": {
+			addr: testAddrWithoutPort,
+			setup: func() dialContextWithRoundRobinDNS {
+				mdc = newMockDialContext([]string{testAddrWithoutPort})
+				return dialContextWithRoundRobinDNS{
+					dialContext: mdc.dialContext,
+					resolver:    &mockedLookupHost{withErr: false},
+				}
```

Comment:
Worth extracting this part to a function?  It is repeated 3 times with slight differences, so pulling those into a single function-call might make it easier to follow.